src/de/dkfz/b080/co/common/WorkflowUsingMergedBams.groovy

```diff
@@ -164,39 +167,57 @@ abstract class WorkflowUsingMergedBams extends Workflow {
             return false
         }
 
-        if (isNoControlWorkflow()) {
-            boolean foundAll = true
-            initialBamFiles.each { BasicBamFile it -> foundAll &= it && ((COFileStageSettings) it.getFileStage()).sample.sampleType == Sample.SampleType.TUMOR }
+        if (isControlWorkflow() && initialBamFiles.size()<2) {
+            context.addError(ExecutionContextError.EXECUTION_NOINPUTDATA.expand("Bam file list is too small, need one bam for no control and two bams for control workflows."))
+            return false
+        }
 
-            if (!foundAll) {
-                context.addErrorEntry(ExecutionContextError.EXECUTION_NOINPUTDATA.expand("Not all found files are tumor BAM files."))
-                return false
-            }
+        // Check, if the bams were created properly
+        if (initialBamFiles.any { it == null }) {
+            context.addError(ExecutionContextError.EXECUTION_NOINPUTDATA.expand("The list of bam files contains an empty entry."))
+            return false
+        }
 
-        } else {
-            boolean foundAll = true
-            BasicBamFile bamControlMerged = initialBamFiles[0]
+        if (initialBamFiles.any { BasicBamFile bam -> bam.fileStage == null }) {
+            context.addError(ExecutionContextError.EXECUTION_NOINPUTDATA.expand("The list of bam files contains an entry with unset sample."))
+            return false
+        }
 
-            for (int i = 1; i < initialBamFiles.size(); i++) {
-                if (initialBamFiles[i] == null || !((COFileStageSettings) initialBamFiles[i].getFileStage()).sample.sampleType == Sample.SampleType.TUMOR) {
-                    context.addErrorEntry(ExecutionContextError.EXECUTION_NOINPUTDATA.expand("Tumor BAM is missing."))
-                    foundAll = false
-                }
-            }
+        // Check, if samples are ok (one control, several tumors)
+        // Control first, if it is a control based workflow, then tumors
 
-            if (bamControlMerged == null || ((COFileStageSettings) bamControlMerged.getFileStage()).sample.sampleType == Sample.SampleType.TUMOR) {
-                context.addErrorEntry(ExecutionContextError.EXECUTION_NOINPUTDATA.expand("Control BAM is missing and workflow is not set to accept tumor only ($IS_NO_CONTROL_WORKFLOW variable)."))
-                foundAll = false
+        List<ExecutionContextError> errors = []
+        int tumorIndex
+        if (isControlWorkflow()) {
+            tumorIndex = 1
+            if (initialBamFiles[0].sample.type != CONTROL) {
+                errors << ExecutionContextError.EXECUTION_NOINPUTDATA.expand("Control bam file is missing and workflow is not set to accept tumor only.")
             }
+        } else {
+            tumorIndex = 0
+        }
 
-            if (!foundAll) return false
+        // Now check tumor files
+
+        def tumorFiles = initialBamFiles[tumorIndex..-1]
+        def tumorFileCount = tumorFiles.size()
```

Comment:
Really not necessary this variable. You just trade 2 times 2 symbols (parentheses)  against "def " which is also 4 symbols and the code does rather get more complex.